sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala

```diff
@@ -265,3 +271,258 @@ private[aggregate] object CollectTopK {
+    if (buffer.nonEmpty) {
+      val sortedBufferWithoutNulls = sortBuffer(buffer)
+      concatSkippingNulls(sortedBufferWithoutNulls)
+    } else {
+      null
+    }
+  }
+
+  private[this] def sortBuffer(buffer: mutable.ArrayBuffer[Any]): mutable.ArrayBuffer[Any] = {
+    if (!orderingFilled) {
+      return buffer
+    }
+    if (noNeedSaveOrderValue) {
+      val ascendingOrdering = PhysicalDataType.ordering(orderExpressions.head.dataType)
+      val ordering =
+        if (orderExpressions.head.direction == Ascending) ascendingOrdering
+        else ascendingOrdering.reverse
+      buffer.sorted(ordering)
+    } else {
+      buffer
+        .asInstanceOf[mutable.ArrayBuffer[InternalRow]]
+        .sorted(bufferOrdering)
+        // drop order values after sort
+        .map(_.get(0, child.dataType))
+    }
+  }
+
+  private[this] def bufferOrdering: Ordering[InternalRow] = {
+    val bufferSortOrder = orderExpressions.zipWithIndex.map {
+      case (originalOrder, i) =>
+        originalOrder.copy(
+          // first value is the evaluated child so add +1 for order's values
+          child = BoundReference(i + 1, originalOrder.dataType, originalOrder.child.nullable)
+        )
+    }
+    new InterpretedOrdering(bufferSortOrder)
+  }
+  private[this] def concatSkippingNulls(buffer: mutable.ArrayBuffer[Any]): Any = {
+    val delimiterValue = getDelimiterValue
+    dataType match {
+      case BinaryType =>
+        val inputs = buffer.filter(_ != null).map(_.asInstanceOf[Array[Byte]])
+        ByteArray.concatWS(delimiterValue.asInstanceOf[Array[Byte]], inputs.toSeq: _*)
+      case _: StringType =>
+        val inputs = buffer.filter(_ != null).map(_.asInstanceOf[UTF8String])
+        UTF8String.fromString(inputs.mkString(delimiterValue.toString))
+    }
+  }
+
+  override def dataType: DataType = child.dataType
+
+  private[this] def getDelimiterValue: Any = {
+    val delimiterValue = delimiter.eval()
+    if (delimiterValue == null) {
+      // default delimiter value
+      dataType match {
+        case _: StringType => UTF8String.fromString("")
+        case _: BinaryType => ByteArray.EMPTY_BYTE
+      }
+    } else {
+      delimiterValue
+    }
+  }
+
+  override def update(buffer: ArrayBuffer[Any], input: InternalRow): ArrayBuffer[Any] = {
+    val value = child.eval(input)
+    if (value != null) {
+      val v = if (noNeedSaveOrderValue) {
+        convertToBufferElement(value)
+      } else {
+        InternalRow.fromSeq(convertToBufferElement(value) +: evalOrderValues(input))
+      }
+      buffer += v
+    }
+    buffer
+  }
+
+  private[this] def evalOrderValues(internalRow: InternalRow): Seq[Any] = {
+    orderExpressions.map(order => convertToBufferElement(order.child.eval(internalRow)))
+  }
+
+  override protected def convertToBufferElement(value: Any): Any = InternalRow.copyValue(value)
+
+  override def children: Seq[Expression] = child +: delimiter +: orderExpressions.map(_.child)
```

Comment:
```suggestion
  override def children: Seq[Expression] = child +: delimiter +: orderExpressions
```

Reply:
The problem here is that `SortOrder` is a very specific expression and exposing it in expression tree "as is" creates a lot of problems:
1. it's `Unevaluable` and according to doc `These expressions don't live past analysis or optimization time` that will be not true for orderExpressions in `Listagg`
2. It breaks distinct and window functions logic with substitution with aliases 

It's probably can be fixed but I don't see a good reason to change the current simple solution

Reply:
I think the doc is a bit inaccurate. `WindowSpecDefinition` also contains `SortOrder` as its child.

The current code is just broken. If you look into `TreeNode` a bit more, the way catalyst framework works is: go through all the case class parameters, for each parameter, if `children` contains it, then we transform this parameter as well. Now we treat `SortOrder`'s child as the child of `ListAgg`, I don't think we can ever transform into `ListAgg#orderExpressions`.

Reply:
Fixed.
But I still don't get it. Why `children` + `withNewChildrenInternal` aren't enough? For what dark magic does the framework need to access the default constructor and all fields?